### ISSUE TITLE
Tracing: Lua `httpCall` requests are sampled regardless of the sampling decision for the overall request

### ISSUE THREAD

**Issue by isker** [envoyproxy/envoy#33164]
*Title*: Lua [`httpCall`](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/lua_filter#httpcall) requests are sampled regardless of the sampling decision for the overall request

*Description*:
If you have a tracer configured on your connection manager and make any `httpCall`s from a Lua filter, spans from the resulting call are sampled by default and are sent to the tracing provider regardless of whether the downstream request itself is to be sampled, e.g. as a result of `random_sampling`. This results in orphaned spans if the downstream request is not being sampled; the `httpCall` span is sent but no other spans from the trace are.

You can disable sampling for the `httpCall` span using the `trace_sampled` option, but not in a way that comports with `random_sampling`, as (as far as I know) there is no way to determine that sampling decision in Lua. (Perhaps you can write a native filter that projects the trace context's `sampled` flag into dynamic metadata, but that is a high bar for users to clear.)

I think the default value of `httpCall`'s [`trace_sampled` flag](https://github.com/envoyproxy/envoy/blob/7f6df241c61094c6e1ec26c8d1acf05f5803a32c/source/extensions/filters/http/lua/lua_filter.cc#L54-L58) should change from [`true`](https://github.com/envoyproxy/envoy/blob/7f6df241c61094c6e1ec26c8d1acf05f5803a32c/envoy/http/async_client.h#L426-L427), to one derived from the trace context for the request.

*Repro steps*:
I will make a minimal reproduction when I have time, hopefully within the next few hours; sorry.

**Comment by isker** [comment]
cc @wbpcode as [you added](https://github.com/envoyproxy/envoy/pull/22871) the `trace_sampled` flag in the first place.

**Comment by wbpcode** [comment]
It's hard to change the default tracing decision (historical reason). May be we can add third state to the `trace_sampled`. If it's nil, then just reuse their parent's tracing decision?

**Comment by isker** [comment]
In the use case in my reproduction, the default is producing strictly invalid results (envoy emits only some spans from the trace; the zipkin collector is waiting to receive the parent span of the `httpCall`, never receives it, and eventually gives up and marks the trace as invalid).

Is there another use case in which the default is not producing strictly invalid results? Why should users have to opt in to getting correct behavior?

Beyond the incorrect behavior of the default, it's an operational footgun: using `httpCall`, envoy will emit substantially more spans (and thus send substantially more HTTP traffic to the collector) than you would expect from your `random_sampling` percentage, if that percentage is low.

**Comment by wbpcode** [comment]
It's hard to follow the actual reason about why the `RequestOptions` setting a default `true` sampling flag. But it exists there for a long time.
May be you can create a PR to change it (with a runtime guard). Then let maintainers from different companies to review it.
If we can finally agree on something, it would be great.

**Comment by wbpcode** [comment]
Yeah. I think you are right.
